ns for fx: add allowlist for ops with same signature across dtypes

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55918 ns for fx: fill out coverage for node I/O types
* #55859 ns for fx: add category for ops which accept fp32 or int8 input
* #55858 ns for fx: move op dtype category mapping to separate file
* #55837 ns for fx: add test for op relationship coverage
* #55805 ns for fx: move pattern utils to separate file
* #55803 ns for fx: make get_reversed_fusions reuse quantization fusions
* #55614 ns for fx: add qat tests cases for shadowed activations
* #55508 ns for fx: qat test cases for unshadowed activations
* #55507 ns for fx: extend functional weight extraction testing to QAT
* #55506 ns for fx: add qat handling for weight extraction
* #55505 ns for fx: make NSTracer inherit from QuantizationTracer
* #55434 ns for fx: change node I/O determination to strict allowlist
* #55433 ns for fx: add functional conv-relu fusion support
* #55432 ns for fx: add test case for linear dynamic
* #55431 ns for fx: fix bug in weight extraction testing
* #55327 ns for fx: add conv1d weight extraction
* #55311 fx quant: fix prepacking for F.conv1d
* #55288 ns for fx: move more weight matching logic to weight_utils.py
* #55287 ns for fx: add F.conv2d and F.conv3d weight extraction
* #55155 ns for fx: rename SugraphTypeRelationship to SubgraphTypeRelationship
* **#55154 ns for fx: add allowlist for ops with same signature across dtypes**
* #55080 ns for fx: add linear-relu mod weight extraction

Summary:

Adds functionality to NS to allow matching nodes which have the
same signature across dtypes.  For now, only the skeleton is added,
we can fill out the rest of the ops later.  This is to unblock
the work to change `cat` to have the same signature for fp32 and int8,
and keep the testing we have for `cat` in NS.

For context, the main reason we are not matching nodes with equal types,
for now, is user defined types for which we do not know the signature.
For now, the design is strictly allowlist of everything. In the future,
we may adjust the design to safely match user defined types.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_ops_with_same_fp32_and_int8_signature
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27504624](https://our.internmc.facebook.com/intern/diff/D27504624)

vkuzo committed:
ns for fx: add allowlist for ops with same signature across dtypes

Summary:

Adds functionality to NS to allow matching nodes which have the
same signature across dtypes.  For now, only the skeleton is added,
we can fill out the rest of the ops later.  This is to unblock
the work to change `cat` to have the same signature for fp32 and int8,
and keep the testing we have for `cat` in NS.

For context, the main reason we are not matching nodes with equal types,
for now, is user defined types for which we do not know the signature.
For now, the design is strictly allowlist of everything. In the future,
we may adjust the design to safely match user defined types.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_ops_with_same_fp32_and_int8_signature
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add allowlist for ops with same signature across dtypes"


Summary:

Adds functionality to NS to allow matching nodes which have the
same signature across dtypes.  For now, only the skeleton is added,
we can fill out the rest of the ops later.  This is to unblock
the work to change `cat` to have the same signature for fp32 and int8,
and keep the testing we have for `cat` in NS.

For context, the main reason we are not matching nodes with equal types,
for now, is user defined types for which we do not know the signature.
For now, the design is strictly allowlist of everything. In the future,
we may adjust the design to safely match user defined types.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_ops_with_same_fp32_and_int8_signature
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27504624](https://our.internmc.facebook.com/intern/diff/D27504624)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add allowlist for ops with same signature across dtypes"


Summary:

Adds functionality to NS to allow matching nodes which have the
same signature across dtypes.  For now, only the skeleton is added,
we can fill out the rest of the ops later.  This is to unblock
the work to change `cat` to have the same signature for fp32 and int8,
and keep the testing we have for `cat` in NS.

For context, the main reason we are not matching nodes with equal types,
for now, is user defined types for which we do not know the signature.
For now, the design is strictly allowlist of everything. In the future,
we may adjust the design to safely match user defined types.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_ops_with_same_fp32_and_int8_signature
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27504624](https://our.internmc.facebook.com/intern/diff/D27504624)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/ns/graph_matcher.py:
```diff
@@ -352,14 +360,30 @@ class GraphMatchingException(Exception):
 class SugraphTypeRelationship(enum.Enum):
     # same type
-    # example: F.linear and toq.linear, or nn.Conv2d and nn.Conv2d
+    # example: F.linear and F.linear, or nn.Conv2d and nn.Conv2d
```
Is this a typo? SugraphType?

vkuzo replied:
yes, the next PR in the stack fixes this

vkuzo committed:
Update on "ns for fx: add allowlist for ops with same signature across dtypes"


Summary:

Adds functionality to NS to allow matching nodes which have the
same signature across dtypes.  For now, only the skeleton is added,
we can fill out the rest of the ops later.  This is to unblock
the work to change `cat` to have the same signature for fp32 and int8,
and keep the testing we have for `cat` in NS.

For context, the main reason we are not matching nodes with equal types,
for now, is user defined types for which we do not know the signature.
For now, the design is strictly allowlist of everything. In the future,
we may adjust the design to safely match user defined types.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_ops_with_same_fp32_and_int8_signature
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27504624](https://our.internmc.facebook.com/intern/diff/D27504624)

[ghstack-poisoned]

vkuzo committed:
rebase on "ns for fx: add allowlist for ops with same signature across dtypes"


Summary:

Adds functionality to NS to allow matching nodes which have the
same signature across dtypes.  For now, only the skeleton is added,
we can fill out the rest of the ops later.  This is to unblock
the work to change `cat` to have the same signature for fp32 and int8,
and keep the testing we have for `cat` in NS.

For context, the main reason we are not matching nodes with equal types,
for now, is user defined types for which we do not know the signature.
For now, the design is strictly allowlist of everything. In the future,
we may adjust the design to safely match user defined types.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_ops_with_same_fp32_and_int8_signature
python test/test_quantization.py TestFXGraphMatcher.test_nodes_with_equal_types_do_not_get_matched
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27504624](https://our.internmc.facebook.com/intern/diff/D27504624)

[ghstack-poisoned]
